### PR TITLE
Addressbook sharing

### PR DESCRIPTION
**Pull request by tomneedham**
TODO:
- [x] Interpret shares on load of the AddressBook model
- [x] Tweak some css in the UI to make it pretty

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
ouch - looks :boom:

**Comment by tomneedham** [comment]
some squashing needed... ;)

**Comment by tomneedham** [comment]
@DeepDiver1975 Broken for you?

**Comment by DeepDiver1975** [comment]
Looks like a rebase is needed ... Toooo many commits

**Comment by tomneedham** [comment]
Mmm clean now :)

File: bower.json [hunk 1]
```diff
@@ -22,6 +22,9 @@
     "angular-uuid4": "~0.3.1",
     "angular-cache": "~4.5.0",
     "angular-route": "^1.5.0",
-    "vcard": "~0.2.3"
+    "vcard": "~0.2.3",
+    "dav": "git@github.com:irgendwie/dav.git#d33ecad1e64202ec5b35b6911708d1253064b5a7",
+    "vcard": "~0.2.3",
```
Comment (Henni) [inline bower.json]: We removed the dav library from bower in one of the last PRs.
Additionally vcard seems to be twice in this list.

File: js/services/addressBook_service.js [hunk 2]
```diff
@@ -72,7 +72,111 @@ app.factory('AddressBookService', ['DavClient', 'DavService', 'SettingsService',
...
+			oSet.appendChild(oSummary);
+
+			if (writable) {
+				var oRW = xmlDoc.createElement('o:read-write');
+				oSet.appendChild(oRW);
+			}
+
+			var body = oShare.outerHTML;
+
+			DavClient.xhr.send(
```
Comment (Henni) [inline js/services/addressBook_service.js]: I believe this is missing a `return`. I'm not sure if https://github.com/owncloud/contactsrework/pull/18/files#diff-dbaa024398385bf6cffe22c422471ff0R96 works without this.

File: js/public/script.js [hunk 3]
```diff
@@ -163,12 +264,66 @@ app.factory('AddressBook', function()
...
+		}
+
+		var owner = props['{DAV:}owner'];
+		if (typeof owner !== 'undefined' && owner.length !== 0) {
+			owner = owner[0].textContent.slice(0, -1);
+			if (owner.startsWith('/remote.php/dav/principals/users/')) {
+				this._properties.owner = owner.substr(33);
+			}
+		}
+		*/
```
Comment (Henni) [inline js/public/script.js]: I know that this is still work in progress... But just a short reminder to remove this code comment block before merging.

File: templates/main.php [hunk 4]
```diff
@@ -12,8 +12,13 @@
 // compiled version of app javascript
 script('contactsrework', 'public/script');
 
+script('contactsrework', 'vendor/angular-bootstrap/ui-bootstrap.min');
+script('contactsrework', 'vendor/angular-bootstrap/ui-bootstrap-tpls.min');
+
 // all styles
 style('contactsrework', 'public/style');
+style('contactsrework', 'addressbooklist');
```
Comment (Henni) [inline templates/main.php]: Please move the styles to `public/style.css`.
We decided to keep all styles in a single file and remove any modularization or build steps for now.

File: js/public/script.js (hunk 3 above)
Comment (tomneedham) [inline js/public/script.js]: Yup this is my thinking out loud of how we interpret shares on the initial PROPFIND. @DeepDiver1975 is implementing it in the dav library tomorrow and then we can use this code.

File: templates/main.php (hunk 4 above)
Comment (tomneedham) [inline templates/main.php]: Cool shall do. Again this was mainly whilst stuff got worked out.

**Comment by DeepDiver1975** [comment]
@Henni @tomneedham @jancborchardt sharees are now read from the propfind response - have fun!

File: js/models/addressBook_model.js [hunk 5]
```diff
@@ -13,9 +13,56 @@ app.factory('AddressBook', function()
...
+
+			sharedWith: {
+				users: [],
+				groups: []
 			}
 
 		});
 		angular.extend(this, data);
+		angular.extend(this, {
+			owner: data.url.split('/').slice(-3, -2)[0]
```
Comment (tomneedham) [inline js/models/addressBook_model.js]: @DeepDiver1975 Is this ok (with regards to how I have found the owner?). Calendar uses the share prop that you just implemented, is that safer?

**Comment by tomneedham** [comment]
@DeepDiver1975 We get a load of lines in the logs that 'principals' sharebackend is not found because of https://github.com/owncloud/core/blob/master/apps/files_sharing/api/sharees.php#L445. Should we add a dummy backend to return no to this and keep the log quiet?

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 We get a load of lines in the logs that 'principals' sharebackend is not found because of https://github.com/owncloud/core/blob/master/apps/files_sharing/api/sharees.php#L445. Should we add a dummy backend to return no to this and keep the log quiet?

or we simply use file for now ....

**Comment by DeepDiver1975** [comment]
I'll rebased this beast ....

File: js/models/addressBook_model.js (hunk 5 above)
Comment (DeepDiver1975) [inline js/models/addressBook_model.js]: I have to double check this ...

Actually I wanted to ask you what for the owner is used.
Best would be to get it from the profind response ...

**Comment by jancborchardt** [comment]
Ok … not sure how I can test this. The UI doesn’t seem to change at all, and I assume there should be something in the settings area?

**Comment by DeepDiver1975** [comment]
Sharing by admin:
bildschirmfoto von 2016-02-24 15-54-49
bildschirmfoto von 2016-02-24 15-55-24
bildschirmfoto von 2016-02-24 15-55-33


As user01:
bildschirmfoto von 2016-02-24 15-56-03

**Comment by jancborchardt** [comment]
How do I get from screenshot 1 to 2? I assume clicking the address book – doesn’t work though here.

Just to be sure, the element looks like this:
```
<li class="ng-scope ng-binding ng-isolate-scope"
  ng-repeat="addressBook in ctrl.addressBooks"
  addressbook="" data="addressBook">
    testAddr
</li>
```

**Comment by DeepDiver1975** [comment]
> How do I get from screenshot 1 to 2? I assume clicking the address book – doesn’t work though here.

there is a share icon (which is not in the screenshot) ... just a sec

**Comment by DeepDiver1975** [comment]
bildschirmvideo von 24 02 2016 16_37_59

**Comment by jancborchardt** [comment]
For the record: doesn’t work with Firefox, only Chrome.

**Comment by DeepDiver1975** [comment]
> For the record: doesn’t work with Firefox, only Chrome.

@tpn please take care - THX

**Comment by DeepDiver1975** [comment]
sharing works - the shared addressbook appears on the sharee - but no contacts are loaded - I suggest to take care of this in a followup pr

**Comment by DeepDiver1975** [comment]
> but no contacts are loaded

this is because of the enabled logic - https://github.com/owncloud/contactsrework/blob/master/js/services/addressBook_service.js#L35

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
let get this in @jancborchardt

File: js/models/addressBook_model.js (hunk 5 above)
Comment (tomneedham) [inline js/models/addressBook_model.js]: Yup this is how calednar does it: https://github.com/owncloud/calendar-rework/blob/master/js/app/models/calendarmodel.js#L107

**Comment by tomneedham** [comment]
> For the record: doesn’t work with Firefox, only Chrome.

Yup I see this too, created https://github.com/owncloud/contactsrework/issues/54

**Comment by jancborchardt** [comment]
I tested this in Chrome: I do get an address book for my other user called `testAddr(jan)` but nothing happens … no contacts displayed by default, nothing changes when I click on »All« or on that address book. cc @DeepDiver1975 
Maybe best to merge anyway and fix stuff going forward?

@tomneedham some additional things to fix, although in a future PR:
- [ ] clicking on the share icon should not only show the share box, but also focus it with the cursor
- [ ] the suggestion dropdown should overlay the content below, not move the whole thing up like now
- [ ] remove the dash behind the name in the list

**Comment by DeepDiver1975** [comment]
> @tomneedham some additional things to fix, although in a future PR:
> 
>     clicking on the share icon should not only show the share box, but also focus it with the cursor
>     the suggestion dropdown should overlay the content below, not move the whole thing up like now
>     remove the dash behind the name in the list

let's add this to #54

**Pull request merged**